### ISSUE TITLE
[Feat] Merge `updateTriggers` and other compound properties when multiple prop objects are supplied to layer

### ISSUE THREAD

**Issue by mz8i** [visgl/deck.gl#6619]
### Target Use Case

The ability to pass multiple prop objects to a layer enables useful patterns where various aspects of the layer are defined by different pieces of code.

This mechanism is currently documented (and works) as having the semantics of `Object.assign()` - merging multiple objects and overwriting duplicate properties.

However, this is a problem when relying on complex properties like `updateTriggers`, `transitions` etc which are nested objects with multiple properties referenced. Consider the following example:

```js
function selectionStyle(selectedFeatureId) {
    return {
        getColor: x => (x.id === selectedFeatureId ? [0, 255, 255] : [100, 100, 100]),
        updateTriggers: {
            getColor: [selectedFeatureId],
        },
    };
}

function typeFilter(selectedType) {
    return {
        getFilterValue: x => (x.properties.type === selectedType ? 1 : 0),
        filterRange: [1, 1],
        extensions: [new DataFilterExtension({ filterSize: 1 })],
        updateTriggers: {
            getFilterValue: [selectedType],
        },
    };
}

const DATA = [ /* ... */ ];

function fooLayer(selectedFeatureId, selectedType) {
    return new GeoJsonLayer(
        {
            id: 'foo',
            data: DATA,
            filled: true,
            stroked: true,
        },
        selectionStyle(selectedFeatureId),
        typeFilter(selectedType)
    );
}
```
This effectively prevents the user from relying on the prop merging mechanism. The user needs to merge the objects manually to avoid losing the relevant fields in `updateTriggers` etc.

### Proposal

Establish a new semantic for merging multiple prop objects supplied to a layer constructor, where complex properties are merged instead of overwritten. This would enable users to write more composable code.

The key props I've identified where merging instead of overwriting would be useful are:
- `updateTriggers`
- `transitions`

For the above example, the expected result would be to end up with
```js
{
//...
updateTriggers: {
   getColor: [selectedFeatureId],
   getFilterValue: [selectedType],
}
//...
}
```
If the `updateTriggers` objects being merged contain duplicate keys, the latest value for the duplicate key should overwrite the earlier ones. This would be consistent with how the simple top-level properties like `getColor` are being overwritten.

There's also `extensions` which is an array. The most desirable semantic here is less clear to me. For most cases, concatenating the arrays would make most sense, but I am not sure what happens if you end up with an `extensions` array containing two extensions of the same type, for example `extensions: [new DataFilterExtension({filterSize: 1}), new DataFilterExtension({filterSize: 2})]`

**Comment by mz8i** [comment]
@Pessimistress @ibgreen I'm wondering if you've had a chance to review this proposal? I appreciate it could be seen as a fairly large change to the semantics of prop merging, but I still believe it would greatly improve the ability for users of deck.gl to write code with better separation of concerns, without having to go outside of vanilla deck.gl to handle merging of `updateTriggers` etc manually

**Comment by Pessimistress** [comment]
My main concern is that this can slow down the prop creation process quite a lot. Even if no merging is required, we will need to create a new object for transitions/updateTriggers and make a full copy. Currently prop merging perf is at >1M iterations/second. There is not a lot of slack for doing extra traversals.

May I ask why you don't want to do it yourself?

**Comment by mz8i** [comment]
Thanks for looking into this. Here's a bit of code which I hope will demonstrate my use case on less of a toy problem than the initial example. I have a project that mostly uses a set number of layers (MVT/GeoJSON and raster tiles) but there are a lot of combinations of data sources / styles / interactive behaviours for the layers. I've tried to partition the bits of functionality so that they can be recombined in any number of ways, while avoiding the subclassing of built-in layer classes if the same functionality can be achieved by just constructing the supported props in the right manner.

Here's an example layer I've created to build an MVTLayer where one of the loaded features can be selected (say, upon clicking). I'm omitting quite a bit of code but hopefully everything is fairly self-explanatory.

```ts
interface SelectableMvtLayerOptions {
  selectionOptions: TileSelectionLayerOptions;
}

export function selectableMvtLayer(
  { selectionOptions }: SelectableMvtLayerOptions,
  ...props
) {
  return mvtLayer(
    {
      binary: false,
      autoHighlight: true,
      highlightColor: [0, 255, 255, 255],
      refinementStrategy: 'best-available',
      renderSubLayers: (tileProps) => [
        geoJsonLayer(tileProps),
        tileSelectionLayer(tileProps, selectionOptions),
      ],
      updateTriggers: {
        renderSubLayers: [selectionOptions.selectedFeatureId],
      },
    },
    props,
  );
}
```
I can use this helper to create an MVT layer with the selection mechanism, but all other props related to style, data etc are not known inside this function. They are just passed onto the MVT layer.
Because the other packaged bits of functionality could rely on `updateTriggers`, we need to merge them according to the proposed semantics. Because there's currently no support for that in vanilla deck.gl, we have two options:
- remember to always merge the props in any place where a deck.gl layer is created. This is fine but creates a risk of someone forgetting to do that which would then introduce unexpected bugs in the app
- wrap vanilla deck.gl layers to handle this automatically and only use the wrapped version. You'll notice the `mvtLayer` function used above and I went down this path so far, the main problem is that then you need to wrap any vanilla layer that you'd like to use, and any third-party layer, and your own subclassed layers.

Here's the code where I do the wrapping:
```ts
/**
 * A set of strategies for merging compound layer properties
 * specific to Deck.GL
 */
const deckPropsMergeStrategies = {
  // merge updateTriggers objects with Object.assign() semantics
  updateTriggers: mergeValue,

  // concatenate extensions arrays
  extensions: appendValue,
};

/**
 * A function to merge multiple props objects passed to a Deck.GL layer.
 * This extends the base Deck.GL behaviour in a few ways:
 * - falsy elements of the array are ignored
 * - nested arrays are flattened
 * - compound props are merged instead of overwritten
 */
function mergeDeckProps(...props: ConfigTree<object>): any {
  const flattenedProps = flattenConfig(props);

  return mergeObjects(flattenedProps, deckPropsMergeStrategies);
}

function wrap(deckClass) {
  return (...props) => new deckClass(mergeDeckProps(props));
}

export const mvtLayer = wrap(MVTLayer);
export const tileLayer = wrap(TileLayer);
export const bitmapLayer = wrap(BitmapLayer);
export const geoJsonLayer = wrap(GeoJsonLayer);
// ...
```

The main reason I am asking for this feature is that it seemed that it'd be most appropriate to support this inside deck.gl rather than invent a whole library around it, package it or recreate it in all projects...

I see your point about performance - could this perhaps be made an opt-in feature, e.g. something like

```jsx
<DeckGL
 // ...
propMergeStrategies={{
    updateTriggers: mergeValue,
    extensions: appendValue
}}
/>
```
where the default is something like `overrideValue`, and the configuration would then be passed all the way down to https://github.com/visgl/deck.gl/blob/master/modules/core/src/lifecycle/create-props.ts#L50 ? This way any potential performance hit doesn't affect people who don't use it.

As a side note, you'll note in the above code that I'm also applying to props the same logic that deck.gl already applies to layers - merging nested arrays of props and ignoring falsy values. This has also proved very useful and would be another feature request (with less of a performance impact I gather) but I didn't want to make two issues at once :)

**Comment by Pessimistress** [comment]
I'm not sure I fully understand your solution. Can't you just replace `tileSelectionLayer(tileProps, selectionOptions)` with `tileSelectionLayer(mergeProps(tileProps, selectionOptions))`?

**Comment by mz8i** [comment]
Actually `tileSelectionLayer` and `selectableMvtLayer` were in the example just to demonstrate the usage, but they have slightly different signatures because they are packaged bits of functionality and therefore at least one parameter is not a deck.gl props object. Should have made that clearer, sorry. I'd mostly focus on the other functions that are wrapped layers, such as `mvtLayer`, `geoJsonLayer`, `bitmapLayer` etc.

But in principle - yes, instead of having the wrapping code and calling `mvtLayer(...)` you could just write `new MVTLayer(mergeProps(...))` everywhere in your app.

This feature request is purely about improving the DX of using deck.gl to build complex projects. The basic semantics of merging props already feel intuitive. But when you use `updateTriggers` or `transitions`, the semantics actually feel counterintuitive, because these props are essentially "bags" of potentially unrelated keys and simply overwriting the whole object doesn't fit the mental model. I understand this is done for performance and consistency, though you could argue that `getFillColor` and `updateTriggers` are props of a different *level*, so they should also be merged on a different *level*...

I think there are two main benefits of adding an (opt-in) feature to specify if the compound props should be overwritten or merged. Both of them solve some problems with the workarounds I mentioned above.
- If someone wants to rely on this behavior across their code, there's just one place in the whole app where this needs to be configured: in the place where the `Deck` instance is created. In contrast, when you have to write `mergeProps(...)` everywhere, it's easy to forget and introduce bugs.
- After activating, the same behavior comes out-of-the-box for all layer classes, including third-party layers, because they all inherit the base layer behavior. In contrast, when you have to wrap every new layer you use with the code that calls `mergeProps(...)` for you, you have to do that for every single layer type you'd like to use, and then you need to ensure that you only ever use the wrapped layer and not the vanilla one.

In summary, the proposal is about tweaking the base layer behavior, so the best place to enable that seemed to be inside the base layer code and not in the user code.

Alternatively, if you don't feel it's worth the effort to implement this change for the benefits I'm outlining above, would you be open to reviewing a PR?
